src/model/formula/Formula.Clause.js

```diff
@@ -67,32 +67,43 @@ class Clause extends Formula {
     return undefined;
   }
 
-  isResolvent(cl1, cl2, renaming, unifier){
-    const newCl1 = cl1.substitute(renaming).substitute(unifier);
-    const newCl2 = cl2.substitute(unifier);
-    for (let lit1 of newCl1.lits){
+  * getResolvents(cl1, cl2) {
+    for (let lit1 of cl1.lits){
       const nlit1 = lit1.negation();
-      for (let lit2 of newCl2.lits){
+      for (let lit2 of cl2.lits){
         if (nlit1.equals(lit2)){
-          if (this.equals(new Clause(newCl1.lits.filter(lit => !lit.equals(lit1))
-            .concat(newCl2.lits.filter(lit => !lit.equals(lit2)))))){
-            return true;
-          }
+          yield new Clause(cl1.lits.filter(lit => !lit.equals(lit1))
+          .concat(cl2.lits.filter(lit => !lit.equals(lit2))));
           break;
         }
       }
     }
+  }
+
+  isResolvent(cl1, cl2, renaming, unifier){
+    const newCl1 = cl1.substitute(renaming).substitute(unifier);
+    const newCl2 = cl2.substitute(unifier);
+    for (const resolvent of this.getResolvents(newCl1, newCl2)) {
+      if (this.equals(resolvent)){
+        return true;
+      }
+    }
     return false;
   }
 
+  * getFactors(cl) {
+    for (let i = 0; i < cl.lits.length; i++){
+      yield [new Clause(cl.lits.filter((_, k) => i !== k)), i];
+    }
+  }

```

Comment:
Podobne ako pri rezolvencii, aj tu by interface generátorovej metódy mal byť `cl.getFactors(unifier)` a mala by generovať iba **naozajstné** faktory. Súčasná implementácia generuje klauzuly, ktoré vôbec nemusia byť nie len faktormi, ale ani logickými dôsledkami vstupnej klauzuly.

Faktory môžete pomerne efektívne generovať tak, že si zoberiete `lm = this.substitute(unifier).getLitsMultiset()` a pre každú dvojicu `[lit, n]` z `lm`, kde je `n >= 2` vygenerujete kluzulu tak, že spravíte ďalší prechod cez `lm`, pri ktorom pre `[lit1, n1]` vyrobíte `n1` kópií `lit1`, ak `lit1 != lit`; ale ak `lit1 == lit`, nakopírujete ho `n - 1`-krát. Tento vnútorný prechod by sa mal dať funkcionálne napísať takto:
```javascript
lm.flatMap(([lit1, n1]) => new Array(lit1 == lit ? n1 - 1 : n1).fill(lit1))
```
Vnútorná funkcia vyrába polia s potrebným počtom kópií `lit1`, `flatMap` tieto polia pospája do jedného. Treba z toho samozrejme ešte urobiť novú klauzulu.

Zdá sa mi, že `isFactor` implementovaná pomocou takejto `getFactors` bude zhruba rovnako (ne)efektívna, ako predchádzajúca priama implementácia `isFactor`.